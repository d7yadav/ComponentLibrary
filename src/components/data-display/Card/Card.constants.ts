import type { CardVariant, CardElevation, CardSize, CardOrientation } from './Card.types';

export const CARD_VARIANTS: Record<CardVariant, CardVariant> = {
  elevated: 'elevated',
  outlined: 'outlined',
  filled: 'filled',
  glass: 'glass',
  gradient: 'gradient',
  interactive: 'interactive',
} as const;

export const CARD_ELEVATIONS: Record<CardElevation, CardElevation> = {
  0: 0,
  1: 1,
  2: 2,
  3: 3,
  4: 4,
  6: 6,
  8: 8,
  12: 12,
  16: 16,
  24: 24,
} as const;

export const CARD_SIZES: Record<CardSize, CardSize> = {
  compact: 'compact',
  comfortable: 'comfortable',
  spacious: 'spacious',
} as const;

export const CARD_ORIENTATIONS: Record<CardOrientation, CardOrientation> = {
  vertical: 'vertical',
  horizontal: 'horizontal',
} as const;

export const CARD_SIZE_CONFIGS = {
  compact: {
    padding: '12px',
    headerPadding: '12px 12px 8px',
    contentPadding: '8px 12px',
    actionsPadding: '8px 12px 12px',
    borderRadius: 8,
    gap: 8,
  },
  comfortable: {
    padding: '16px',
    headerPadding: '16px 16px 12px',
    contentPadding: '12px 16px',
    actionsPadding: '12px 16px 16px',
    borderRadius: 12,
    gap: 12,
  },
  spacious: {
    padding: '24px',
    headerPadding: '24px 24px 16px',
    contentPadding: '16px 24px',
    actionsPadding: '16px 24px 24px',
    borderRadius: 16,
    gap: 16,
  },
} as const;

export const CARD_ELEVATION_SHADOWS = {
  0: 'none',
  1: '0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24)',
  2: '0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23)',
  3: '0 10px 20px rgba(0, 0, 0, 0.19), 0 6px 6px rgba(0, 0, 0, 0.23)',
  4: '0 14px 28px rgba(0, 0, 0, 0.25), 0 10px 10px rgba(0, 0, 0, 0.22)',
  6: '0 19px 38px rgba(0, 0, 0, 0.30), 0 15px 12px rgba(0, 0, 0, 0.22)',
  8: '0 25px 50px rgba(0, 0, 0, 0.25)',
  12: '0 35px 70px rgba(0, 0, 0, 0.25)',
  16: '0 45px 90px rgba(0, 0, 0, 0.25)',
  24: '0 60px 120px rgba(0, 0, 0, 0.30)',
} as const;

export const CARD_INTERACTION_STATES = {
  hover: {
    elevationIncrease: 2,
    scaleTransform: 1.02,
    transition: 'all 0.2s cubic-bezier(0.25, 0.46, 0.45, 0.94)',
  },
  active: {
    scaleTransform: 0.98,
    transition: 'all 0.1s cubic-bezier(0.25, 0.46, 0.45, 0.94)',
  },
  focus: {
    outlineWidth: 2,
    outlineOffset: 2,
  },
} as const;

export const CARD_MEDIA_DEFAULTS = {
  height: 200,
  objectFit: 'cover' as const,
  loading: 'lazy' as const,
  borderRadius: {
    top: true,
    bottom: false,
  },
} as const;

export const GLASS_MORPHISM_CARD_CONFIG = {
  backdropFilter: 'blur(12px)',
  backdropFilterFallback: 'blur(8px)',
  borderOpacity: 0.15,
  backgroundOpacity: 0.08,
  saturate: 1.8,
} as const;

export const ACCESSIBILITY_CONSTANTS = {
  minTouchTarget: 44, // Minimum touch target size in pixels
  focusOutlineWidth: 2,
  focusOutlineOffset: 2,
  interactiveRole: 'button',
  cardRole: 'article',
} as const;

export const CARD_ANIMATION_DURATIONS = {
  hover: '0.2s',
  press: '0.1s',
  focus: '0.15s',
  loading: '1.5s',
} as const;

export const CARD_ANIMATION_EASINGS = {
  spring: 'cubic-bezier(0.25, 0.46, 0.45, 0.94)',
  bounce: 'cubic-bezier(0.68, -0.55, 0.265, 1.55)',
  smooth: 'cubic-bezier(0.4, 0, 0.2, 1)',
} as const;