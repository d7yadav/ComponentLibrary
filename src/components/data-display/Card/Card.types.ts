import type { CardProps as MuiCardProps } from '@mui/material/Card';
import type { ReactNode, MouseEvent } from 'react';

export type CardVariant = 
  | 'elevated'
  | 'outlined'
  | 'filled'
  | 'glass'
  | 'gradient'
  | 'interactive';

export type CardElevation = 0 | 1 | 2 | 3 | 4 | 6 | 8 | 12 | 16 | 24;

export type CardSize = 'compact' | 'comfortable' | 'spacious';

export type CardOrientation = 'vertical' | 'horizontal';

export interface CardProps extends Omit<MuiCardProps, 'variant'> {
  variant?: CardVariant;
  
  elevation?: CardElevation;
  
  size?: CardSize;
  
  orientation?: CardOrientation;
  
  interactive?: boolean;
  
  disabled?: boolean;
  
  selected?: boolean;
  
  onClick?: (event: MouseEvent<HTMLDivElement>) => void;
  
  header?: ReactNode;
  
  media?: ReactNode;
  
  children?: ReactNode;
  
  actions?: ReactNode;
  
  className?: string;
  
  'aria-label'?: string;
  
  role?: string;
  
  tabIndex?: number;
}

export interface CardHeaderProps {
  title?: ReactNode;
  
  subtitle?: ReactNode;
  
  avatar?: ReactNode;
  
  action?: ReactNode;
  
  className?: string;
  
  titleTypographyProps?: object;
  
  subtitleTypographyProps?: object;
  
  onClick?: (event: React.MouseEvent<HTMLDivElement>) => void;
  
  'aria-label'?: string;
  
  'aria-describedby'?: string;
  
  disabled?: boolean;
  
  tabIndex?: number;
  
  onKeyDown?: (event: React.KeyboardEvent<HTMLDivElement>) => void;
}

export interface CardContentProps {
  children?: ReactNode;
  
  className?: string;
  
  size?: CardSize;
  
  onClick?: (event: React.MouseEvent<HTMLDivElement>) => void;
  
  disabled?: boolean;
  
  tabIndex?: number;
  
  onKeyDown?: (event: React.KeyboardEvent<HTMLDivElement>) => void;
  
  'aria-label'?: string;
  
  'aria-describedby'?: string;
  
  role?: string;
}

export interface CardMediaProps {
  src?: string;
  
  alt?: string;
  
  component?: 'img' | 'video' | 'div';
  
  height?: number | string;
  
  width?: number | string;
  
  objectFit?: 'cover' | 'contain' | 'fill' | 'scale-down' | 'none';
  
  loading?: 'lazy' | 'eager';
  
  showLoading?: boolean;
  
  className?: string;
  
  children?: ReactNode;
}

export interface CardActionsProps {
  children?: ReactNode;
  
  disableSpacing?: boolean;
  
  className?: string;
  
  align?: 'left' | 'center' | 'right' | 'space-between';
  
  size?: CardSize;
  
  disabled?: boolean;
}

export interface CardStyleProps {
  variant: CardVariant;
  elevation: CardElevation;
  size: CardSize;
  orientation: CardOrientation;
  interactive: boolean;
  disabled: boolean;
  selected: boolean;
  hasHeader: boolean;
  hasMedia: boolean;
  hasActions: boolean;
}