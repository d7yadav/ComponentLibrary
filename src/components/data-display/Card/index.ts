export { Card } from './Card';
export { CardHeader } from './CardHeader';
export { CardContent } from './CardContent';
export { CardMedia } from './CardMedia';
export { CardActions } from './CardActions';

export type { 
  CardProps, 
  CardHeaderProps,
  CardContentProps,
  CardMediaProps,
  CardActionsProps,
  CardVariant, 
  CardElevation,
  CardSize, 
  CardOrientation,
  CardStyleProps 
} from './Card.types';

export {
  CARD_VARIANTS,
  CARD_ELEVATIONS,
  CARD_SIZES,
  CARD_ORIENTATIONS,
  CARD_SIZE_CONFIGS,
  CARD_ELEVATION_SHADOWS,
  CARD_INTERACTION_STATES,
  CARD_MEDIA_DEFAULTS,
  GLASS_MORPHISM_CARD_CONFIG,
  ACCESSIBILITY_CONSTANTS,
  CARD_ANIMATION_DURATIONS,
  CARD_ANIMATION_EASINGS,
} from './Card.constants';