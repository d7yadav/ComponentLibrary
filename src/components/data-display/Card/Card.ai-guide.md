# ü§ñ AI Coding Guide: Card

## Component Overview

**Category**: data-display  
**Complexity**: complex  
**AI Suitability**: High - Well-structured for AI development

## Quick AI Reference

### Props Interface
No specific props defined.

### Available Variants
- `interactive`
- `elevation`

## AI Implementation Guide

### Step-by-Step Implementation

1. **Import the component**:
   ```typescript
   import { Card } from '@/components/data-display/Card';
   ```

2. **Define props interface** (if extending):
   ```typescript
   interface MyCardProps extends CardProps {
     // Additional props
   }
   ```

3. **Implement with proper TypeScript**:
   ```typescript
   const MyCard = (props: MyCardProps) => {
     return <Card {...props} />;
   };
   ```

### AI-Specific Guidelines

- Always use TypeScript strict mode
- Implement proper prop validation
- Include accessibility attributes
- Follow component naming conventions
- Use absolute imports with @ prefix

## Common AI Coding Mistakes

### Mistakes to Avoid

1. **‚ùå Using default exports**
   ```typescript
   // Wrong
   export default Card;
   ```
   
   **‚úÖ Use named exports**
   ```typescript
   // Correct
   export { Card };
   ```

2. **‚ùå Missing TypeScript types**
   ```typescript
   // Wrong
   const props: any = { ... };
   ```
   
   **‚úÖ Proper typing**
   ```typescript
   // Correct
   const props: CardProps = { ... };
   ```

3. **‚ùå Inline styles without theme**
   ```typescript
   // Wrong
   <Card style={{ color: 'red' }} />
   ```
   
   **‚úÖ Theme-based styling**
   ```typescript
   // Correct
   <Card sx={{ color: 'error.main' }} />
   ```

4. **‚ùå Missing accessibility attributes**
   ```typescript
   // Wrong
   <Card onClick={handler} />
   ```
   
   **‚úÖ Proper accessibility**
   ```typescript
   // Correct
   <Card 
     onClick={handler}
     aria-label="Descriptive action"
     role="button"
   />
   ```

## AI Quality Checklist

### Pre-Implementation Checklist

- [ ] Component follows naming conventions (PascalCase)
- [ ] Props interface is properly typed with `CardProps`
- [ ] Component uses named exports only
- [ ] Absolute imports are used (`@/components/...`)
- [ ] Theme integration is implemented
- [ ] Accessibility attributes are included
- [ ] Component is responsive
- [ ] Error boundaries are considered
- [ ] Performance optimizations are applied

### Code Quality Checklist

- [ ] TypeScript strict mode compliance
- [ ] No `any` types used
- [ ] Proper error handling
- [ ] Consistent code formatting
- [ ] JSDoc documentation added
- [ ] Component is testable
- [ ] Storybook stories created
- [ ] Accessibility tested

### Performance Checklist

- [ ] React.memo used if appropriate
- [ ] Expensive operations memoized
- [ ] Bundle size impact considered
- [ ] Re-render optimization implemented
- [ ] Props destructuring optimized

## AI Testing Patterns

### Recommended Testing Approach

1. **Unit Tests**:
   ```typescript
   import { render, screen } from '@testing-library/react';
   import { Card } from './Card';
   
   describe('Card', () => {
     it('renders with default props', () => {
       render(<Card />);
       expect(screen.getByRole('...')).toBeInTheDocument();
     });
   });
   ```

2. **Props Testing**:
   ```typescript
   it('handles all prop combinations', () => {
     const props: CardProps = {
       // Test all required props
     };
     render(<Card {...props} />);
   });
   ```

3. **Accessibility Testing**:
   ```typescript
   import { axe, toHaveNoViolations } from 'jest-axe';
   
   it('meets accessibility standards', async () => {
     const { container } = render(<Card />);
     const results = await axe(container);
     expect(results).toHaveNoViolations();
   });
   ```

4. **Interaction Testing**:
   ```typescript
   import userEvent from '@testing-library/user-event';
   
   it('handles user interactions', async () => {
     const user = userEvent.setup();
     const handleClick = vi.fn();
     render(<Card onClick={handleClick} />);
     
     await user.click(screen.getByRole('...'));
     expect(handleClick).toHaveBeenCalled();
   });
   ```

## AI Recommendations

üß™ **Missing Tests**: Implement comprehensive test suite with >95% coverage
‚ö° **Performance**: Use React.memo and useMemo for optimization
‚ôø **Accessibility**: Ensure WCAG 2.1 AA compliance
üé® **Theming**: Leverage theme system for consistent styling

---

*This guide is optimized for AI-assisted development. Follow these patterns for consistent, high-quality component implementation.*
