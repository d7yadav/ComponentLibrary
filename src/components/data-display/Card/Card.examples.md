# ðŸ“– Card Examples

## Basic Usage

```typescript
import { Card } from '@/components/data-display/Card';

// Basic usage
const Example = () => {
  return (
    <Card>
      Basic Card example
    </Card>
  );
};
```

## Advanced Examples

```typescript
// Advanced configuration
const AdvancedExample = () => {
  const [value, setValue] = useState('');
  
  return (
    <Card
      variant="interactive"
      disabled={false}
      onChange={(e) => setValue(e.target.value)}
      sx={{
        borderRadius: 2,
        '&:hover': {
          transform: 'translateY(-2px)',
        },
      }}
    >
      Advanced Card
    </Card>
  );
};
```

## Integration Examples

```typescript
// Integration with form libraries
import { useController, Control } from 'react-hook-form';

interface FormIntegrationProps {
  name: string;
  control: Control;
}

const FormIntegration = ({ name, control }: FormIntegrationProps) => {
  const { field, fieldState } = useController({
    name,
    control,
    defaultValue: '',
  });

  return (
    <Card
      {...field}
      error={!!fieldState.error}
      helperText={fieldState.error?.message}
    />
  );
};
```

## Edge Cases

```typescript
// Edge case: Empty states
const EmptyState = () => (
  <Card disabled>
    No content available
  </Card>
);

// Edge case: Error states
const ErrorState = () => (
  <Card 
    error
    helperText="Something went wrong"
  />
);

// Edge case: Loading states
const LoadingState = () => (
  <Card 
    loading
    disabled
  />
);
```

## Performance Optimization Examples

```typescript
// Memoized component for performance
const OptimizedCard = React.memo((props: CardProps) => {
  const memoizedValue = useMemo(() => {
    // Expensive calculation
    return heavyComputation(props.value);
  }, [props.value]);

  return (
    <Card 
      {...props}
      value={memoizedValue}
    />
  );
});

// Lazy loading for large lists
const LazyList = () => {
  return (
    <VirtualizedList
      itemCount={1000}
      renderItem={({ index }) => (
        <Card key={index}>
          Item {index}
        </Card>
      )}
    />
  );
};
```

## Accessibility Examples

```typescript
// Full accessibility implementation
const AccessibleExample = () => {
  return (
    <Card
      aria-label="Descriptive label for screen readers"
      aria-describedby="helper-text"
      role="button"
      tabIndex={0}
      onKeyDown={(e) => {
        if (e.key === 'Enter' || e.key === ' ') {
          handleAction();
        }
      }}
    >
      Accessible Card
    </Card>
  );
};

// High contrast mode support
const HighContrastExample = () => {
  return (
    <Card
      sx={{
        '@media (prefers-contrast: high)': {
          borderWidth: 2,
          borderColor: 'text.primary',
        },
      }}
    >
      High contrast support
    </Card>
  );
};
```

---

*These examples demonstrate real-world usage patterns and best practices for the Card component.*
