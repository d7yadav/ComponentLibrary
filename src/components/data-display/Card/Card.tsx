import React, { forwardRef, memo } from 'react';

import { 
  CARD_VARIANTS, 
  CARD_ELEVATIONS, 
  CARD_SIZES, 
  CARD_ORIENTATIONS,
  ACCESSIBILITY_CONSTANTS 
} from './Card.constants';
import { StyledCard } from './Card.styles';
import type { CardProps } from './Card.types';

/**
 * Enhanced Card component with multiple variants and accessibility features
 * 
 * Features:
 * - 6 variants: elevated, outlined, filled, glass, gradient, interactive
 * - Multiple elevation levels (0-24)
 * - Size variants: compact, comfortable, spacious
 * - Orientation: vertical, horizontal
 * - Interactive states with hover effects
 * - Glass morphism and gradient variants
 * - WCAG 2.1 AA accessibility compliance
 * - Dark theme support
 */
export const Card = memo(forwardRef<HTMLDivElement, CardProps>((props, ref): React.ReactElement => {
  const {
    variant = CARD_VARIANTS.elevated,
    elevation = CARD_ELEVATIONS[1],
    size = CARD_SIZES.comfortable,
    orientation = CARD_ORIENTATIONS.vertical,
    interactive = false,
    disabled = false,
    selected = false,
    onClick,
    header,
    media,
    children,
    actions,
    className,
    'aria-label': ariaLabel,
    role,
    tabIndex,
    ...other
  } = props;
  
  // Determine if card should be interactive
  const isInteractive = variant === 'interactive' || interactive || Boolean(onClick);
  
  // Handle click events
  const handleClick = (event: React.MouseEvent<HTMLDivElement>) => {
    if (disabled) {
      event.preventDefault();
      return;
    }
    onClick?.(event);
  };
  
  // Accessibility props
  const accessibilityProps = {
    'aria-label': ariaLabel,
    'aria-disabled': disabled,
    role: role || (isInteractive ? ACCESSIBILITY_CONSTANTS.interactiveRole : ACCESSIBILITY_CONSTANTS.cardRole),
    ...(tabIndex !== undefined && { tabIndex: disabled ? -1 : tabIndex }),
    ...(isInteractive && !disabled && { tabIndex: tabIndex ?? 0 }),
  };

  return (
    <StyledCard
      data-testid="card"
      ref={ref}
      // MUI props (variant/elevation) are set to defaults, custom props control appearance
      variant="elevation"
      elevation={0}
      // Custom props for styled component
      customVariant={variant}
      customElevation={elevation}
      customSize={size}
      customOrientation={orientation}
      interactive={isInteractive}
      disabled={disabled}
      selected={selected}
      {...(className && { className })}
      onClick={isInteractive ? handleClick : undefined}
      {...accessibilityProps}
      {...other}
    >
      {/* Render header, media, children, and actions in order */}
      {header}
      {media}
      {children}
      {actions}
    </StyledCard>
  );
}));

Card.displayName = 'Card';