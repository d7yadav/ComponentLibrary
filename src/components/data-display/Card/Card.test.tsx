/**
 * @fileoverview Card Component Tests
 * @author AI Generated Tests - 2025-07-13
 */

import { ThemeProvider , createTheme } from '@mui/material/styles';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe, toHaveNoViolations } from 'jest-axe';

import { Card } from './Card';
import { CARD_CONSTANTS } from './Card.constants';
import type { CardProps } from './Card.types';

// Add jest-axe matcher
expect.extend(toHaveNoViolations);

// Test theme
const testTheme = createTheme();

// Test wrapper component
const TestWrapper = ({ children }: { children: React.ReactNode }) => (
  <ThemeProvider theme={testTheme}>{children}</ThemeProvider>
);

// Helper function to render component with theme
const renderWithTheme = (ui: React.ReactElement, options = {}) => {
  return render(ui, { wrapper: TestWrapper, ...options });
};

describe('Card Component', () => {
  // Basic rendering tests
  describe('Rendering', () => {
    it('renders without crashing', () => {
      renderWithTheme(<Card>Test content</Card>);
      expect(screen.getByRole('article')).toBeInTheDocument();
          });

    it('renders children correctly', () => {
      const testContent = 'Test Card Content';
      renderWithTheme(<Card>{testContent}</Card>);
      expect(screen.getByText(testContent)).toBeInTheDocument();
    });

    it('renders with custom test id', () => {
      const testId = 'custom-card';
      renderWithTheme(
        <Card data-testid={testId}>Test</Card>
      );
      expect(screen.getByTestId(testId)).toBeInTheDocument();
    });

    it('applies custom className', () => {
      const customClass = 'custom-class';
      renderWithTheme(
        <Card className={customClass}>Test</Card>
      );
      expect(screen.getByRole('article')).toHaveClass(customClass);
    });
  });

  // Props testing
  describe('Props', () => {
    describe('variant prop', () => {
      const variants: CardProps['variant'][] = [
        'elevated',
        'outlined',
        'filled',
        'glass',
        'gradient',
        'interactive',
      ];

      variants.forEach((variant) => {
        it(`renders ${variant} variant correctly`, () => {
          renderWithTheme(
            <Card variant={variant} data-testid="variant-test">
              {variant} variant
            </Card>
          );
          const element = screen.getByTestId('variant-test');
          expect(element).toBeInTheDocument();
        });
      });
    });

    describe('size prop', () => {
      const sizes: CardProps['size'][] = [
        'compact',
        'comfortable',
        'spacious',
      ];

      sizes.forEach((size) => {
        it(`renders ${size} size correctly`, () => {
          renderWithTheme(
            <Card size={size} data-testid="size-test">
              {size} size
            </Card>
          );
          const element = screen.getByTestId('size-test');
          expect(element).toBeInTheDocument();
        });
      });
    });



    it('handles disabled state correctly', () => {
      const handleClick = vi.fn();
      renderWithTheme(
        <Card disabled onClick={handleClick} data-testid="disabled-test">
          Disabled Card
        </Card>
      );
      
      const element = screen.getByTestId('disabled-test');
      expect(element).toHaveAttribute('aria-disabled', 'true');
    });
  });

  // Interaction testing
  describe('Interactions', () => {
    it('handles click events when interactive', async () => {
      const user = userEvent.setup();
      const handleClick = vi.fn();
      
      renderWithTheme(
        <Card 
          interactive 
          onClick={handleClick} 
          data-testid="interactive-card"
        >
          Clickable card content
        </Card>
      );
      
      const card = screen.getByTestId('interactive-card');
      await user.click(card);
      
      expect(handleClick).toHaveBeenCalledTimes(1);
      expect(card).toHaveAttribute('tabIndex', '0');
    });

    it('supports keyboard navigation when interactive', async () => {
      const user = userEvent.setup();
      const handleClick = vi.fn();
      
      renderWithTheme(
        <Card 
          interactive 
          onClick={handleClick} 
          data-testid="keyboard-card"
        >
          Keyboard accessible card
        </Card>
      );
      
      const card = screen.getByTestId('keyboard-card');
      
      // Focus the card
      card.focus();
      expect(card).toHaveFocus();
      
      // Test that card is keyboard accessible
      expect(card).toHaveAttribute('tabIndex', '0');
      expect(card).toHaveAttribute('role', 'button');
    });

    it('does not handle interactions when disabled', async () => {
      const user = userEvent.setup();
      const handleClick = vi.fn();
      
      renderWithTheme(
        <Card 
          interactive 
          disabled 
          onClick={handleClick} 
          data-testid="disabled-card"
        >
          Disabled card
        </Card>
      );
      
      const card = screen.getByTestId('disabled-card');
      
      // Should be marked as disabled
      expect(card).toHaveAttribute('aria-disabled', 'true');
      
      // Click should not work (component prevents it)
      await user.click(card);
      expect(handleClick).not.toHaveBeenCalled();
    });

    it('shows selection state correctly', () => {
      renderWithTheme(
        <Card selected data-testid="selected-card">
          Selected card content
        </Card>
      );
      
      const card = screen.getByTestId('selected-card');
      expect(card).toBeInTheDocument();
      // Selection state is handled via styling props
    });
  });



  // Accessibility testing
  describe('Accessibility', () => {
    it('meets WCAG accessibility guidelines', async () => {
      const { container } = renderWithTheme(
        <Card>Accessible Card</Card>
      );
      
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });

    it('supports screen readers', () => {
      const ariaLabel = 'Screen reader label';
      renderWithTheme(
        <Card aria-label={ariaLabel}>
          Card for screen readers
        </Card>
      );
      
      expect(screen.getByLabelText(ariaLabel)).toBeInTheDocument();
    });

    it('supports keyboard navigation when interactive', () => {
      renderWithTheme(
        <Card interactive tabIndex={0}>
          Keyboard Card
        </Card>
      );
      
      const element = screen.getByRole('button');
      element.focus();
      expect(element).toHaveFocus();
      expect(element).toHaveAttribute('tabIndex', '0');
    });

    it('provides proper ARIA attributes', () => {
      renderWithTheme(
        <Card 
          aria-describedby="description"
          aria-controls="controlled-element"
        >
          ARIA Card
        </Card>
      );
      
      const element = screen.getByRole('article');
      expect(element).toHaveAttribute('aria-describedby', 'description');
      expect(element).toHaveAttribute('aria-controls', 'controlled-element');
    });
  });

  // Theme integration testing
  describe('Theme Integration', () => {
    it('respects theme customizations', () => {
      const customTheme = createTheme({
        components: {
          MuiCard: {
            defaultProps: {
              variant: 'elevated',
            },
            styleOverrides: {
              root: {
                backgroundColor: 'rgb(255, 0, 0)',
              },
            },
          },
        },
      });

      render(
        <ThemeProvider theme={customTheme}>
          <Card data-testid="themed-card">
            Themed Card
          </Card>
        </ThemeProvider>
      );

      const element = screen.getByTestId('themed-card');
      expect(element).toBeInTheDocument();
    });

    it('supports dark mode', () => {
      const darkTheme = createTheme({
        palette: {
          mode: 'dark',
        },
      });

      render(
        <ThemeProvider theme={darkTheme}>
          <Card>Dark Mode Card</Card>
        </ThemeProvider>
      );

      expect(screen.getByRole('article')).toBeInTheDocument();
    });
  });

  // Performance testing
  describe('Performance', () => {
    it('renders efficiently with many instances', () => {
      const manyComponents = Array.from({ length: 100 }, (_, i) => (
        <Card key={i}>Card {i}</Card>
      ));

      const { container } = renderWithTheme(<div>{manyComponents}</div>);
      expect(container.querySelectorAll('[role="article"]')).toHaveLength(100);
    });

    it('respects reduced motion preferences', () => {
      // Mock prefers-reduced-motion
      Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: vi.fn().mockImplementation((query) => ({
          matches: query === '(prefers-reduced-motion: reduce)',
          media: query,
          onchange: null,
          addListener: vi.fn(),
          removeListener: vi.fn(),
          addEventListener: vi.fn(),
          removeEventListener: vi.fn(),
          dispatchEvent: vi.fn(),
        })),
      });

      renderWithTheme(
        <Card animate="false">
          No Animation Card
        </Card>
      );

      expect(screen.getByRole('article')).toBeInTheDocument();
    });
  });

  // Edge cases and error handling
  describe('Edge Cases', () => {
    it('handles empty children gracefully', () => {
      renderWithTheme(<Card />);
      expect(screen.getByRole('article')).toBeInTheDocument();
    });

    it('handles null children gracefully', () => {
      renderWithTheme(<Card>{null}</Card>);
      expect(screen.getByRole('article')).toBeInTheDocument();
    });

    it('handles undefined props gracefully', () => {
      renderWithTheme(
        <Card 
          variant={undefined as any} 
          size={undefined as any}
        >
          Undefined Props
        </Card>
      );
      expect(screen.getByRole('article')).toBeInTheDocument();
    });

    it('handles complex configurations', () => {
      renderWithTheme(
        <Card
          variant="elevated"
          size="compact"
          disabled={false}
          animate={true}
        >
          Complex Card
        </Card>
      );
      
      expect(screen.getByRole('article')).toBeInTheDocument();
    });
  });

  // Snapshot testing
  describe('Snapshots', () => {
    it('matches snapshot for default props', () => {
      const { container } = renderWithTheme(
        <Card>Default Card</Card>
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for all variants', () => {
      const variants: CardProps['variant'][] = [
        'elevated',
        'outlined',
        'filled',
        'glass',
        'gradient',
        'interactive',
      ];

      variants.forEach((variant) => {
        const { container } = renderWithTheme(
          <Card variant={variant}>{variant} Card</Card>
        );
        expect(container.firstChild).toMatchSnapshot(`variant-${variant}`);
      });
    });
  });
});
