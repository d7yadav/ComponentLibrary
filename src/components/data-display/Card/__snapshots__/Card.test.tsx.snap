// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card Component > Snapshots > matches snapshot for all variants > variant-elevated 1`] = `
<div
  aria-disabled="false"
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-daihf1-MuiPaper-root-MuiCard-root"
  data-testid="card"
  role="article"
  style="--Paper-shadow: none;"
>
  elevated
   Card
</div>
`;

exports[`Card Component > Snapshots > matches snapshot for all variants > variant-filled 1`] = `
<div
  aria-disabled="false"
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-x9o42e-MuiPaper-root-MuiCard-root"
  data-testid="card"
  role="article"
  style="--Paper-shadow: none;"
>
  filled
   Card
</div>
`;

exports[`Card Component > Snapshots > matches snapshot for all variants > variant-glass 1`] = `
<div
  aria-disabled="false"
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-10duw0y-MuiPaper-root-MuiCard-root"
  data-testid="card"
  role="article"
  style="--Paper-shadow: none;"
>
  glass
   Card
</div>
`;

exports[`Card Component > Snapshots > matches snapshot for all variants > variant-gradient 1`] = `
<div
  aria-disabled="false"
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-tmc9b2-MuiPaper-root-MuiCard-root"
  data-testid="card"
  role="article"
  style="--Paper-shadow: none;"
>
  gradient
   Card
</div>
`;

exports[`Card Component > Snapshots > matches snapshot for all variants > variant-interactive 1`] = `
<div
  aria-disabled="false"
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-15oix20-MuiPaper-root-MuiCard-root"
  data-testid="card"
  role="button"
  style="--Paper-shadow: none;"
  tabindex="0"
>
  interactive
   Card
</div>
`;

exports[`Card Component > Snapshots > matches snapshot for all variants > variant-outlined 1`] = `
<div
  aria-disabled="false"
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-c0kh2e-MuiPaper-root-MuiCard-root"
  data-testid="card"
  role="article"
  style="--Paper-shadow: none;"
>
  outlined
   Card
</div>
`;

exports[`Card Component > Snapshots > matches snapshot for default props 1`] = `
<div
  aria-disabled="false"
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-daihf1-MuiPaper-root-MuiCard-root"
  data-testid="card"
  role="article"
  style="--Paper-shadow: none;"
>
  Default Card
</div>
`;
